### PR TITLE
custom callback to reduce weight decay

### LINKED ISSUE #1347: Decay the weight decay parameter via a callback function
## What and why
Context https://github.com/aguirre-lab/ml4c3/pull/1334/files/5c07cc0b8ab15f5cb9021c638a0b3d9e133240a5#r573264506

Useful for `AdamW`.

### LINKED ISSUE #1366: connect output and filter tensors to explore
## What and why
in explore, output and filter tensors are not used - but it might be useful to accomplish the filtering functionality of those tensors without cluttering the output of explore

## Solution(s)
connect up the piping here
https://github.com/aguirre-lab/ml4c3/blob/60c6f66ebe9743168a23fd707d4c682268541058/ml4c3/explorations.py#L1023

## Acceptance criteria
below script uses `--filter_tensors` as expected:
```
./scripts/run.sh $PWD/ml4c3/recipes.py explore \
--num_workers 20 \
--tensors     /storage/shared/ecg/mgh \
--tensors     $HOME/dropbox/sts-data/mgh.csv sts \
--samples_per_patient 1 \
--input_tensors \
    sts_death_60_days_after_ecg \
    sts_stroke_60_days_after_ecg \
    sts_renal_failure_60_days_after_ecg \
    sts_prolonged_ventilation_60_days_after_ecg \
    sts_dsw_infection_60_days_after_ecg \
    sts_reoperation_60_days_after_ecg \
    sts_long_stay_60_days_after_ecg \
    ecg_2500_60_days_before_sts \
    ecg_age_60_days_before_sts \
    ecg_sex_60_days_before_sts \
    ecg_rate_md_60_days_before_sts \
    ecg_pr_md_60_days_before_sts \
    ecg_qt_md_60_days_before_sts \
    ecg_qrs_md_60_days_before_sts \
    ecg_raxis_md_60_days_before_sts \
    ecg_paxis_md_60_days_before_sts \
    ecg_taxis_md_60_days_before_sts \
--filter_tensors \
    age_scaled_60_days_after_ecg \
    carshock_60_days_after_ecg \
    chf_60_days_after_ecg \
    chrlungd_60_days_after_ecg \
    classnyh_60_days_after_ecg \
    creatlst_scaled_60_days_after_ecg \
    cva_60_days_after_ecg \
    cvd_60_days_after_ecg \
    cvdpcarsurg_60_days_after_ecg \
    cvdtia_60_days_after_ecg \
    diabetes_60_days_after_ecg \
    dialysis_60_days_after_ecg \
    ethnicity_60_days_after_ecg \
    gender_60_days_after_ecg \
    hct_scaled_60_days_after_ecg \
    hdef_scaled_60_days_after_ecg \
    heightcm_scaled_60_days_after_ecg \
    hypertn_60_days_after_ecg \
    immsupp_60_days_after_ecg \
    incidenc_60_days_after_ecg \
    infendty_60_days_after_ecg \
    medadp5days_60_days_after_ecg \
    medgp_60_days_after_ecg \
    medinotr_60_days_after_ecg \
    medster_60_days_after_ecg \
    numdisv_60_days_after_ecg \
    platelets_scaled_60_days_after_ecg \
    pocpci_60_days_after_ecg \
    pocpciin_60_days_after_ecg \
    prcab_60_days_after_ecg \
    prcvint_60_days_after_ecg \
    prvalve_60_days_after_ecg \
    pvd_60_days_after_ecg \
    raceasian_60_days_after_ecg \
    raceblack_60_days_after_ecg \
    racecaucasian_60_days_after_ecg \
    racenativeam_60_days_after_ecg \
    raceothernativepacific_60_days_after_ecg \
    resusc_60_days_after_ecg \
    status_60_days_after_ecg \
    vdinsufa_60_days_after_ecg \
    vdinsufm_60_days_after_ecg \
    vdinsuft_60_days_after_ecg \
    vdstena_60_days_after_ecg \
    vdstenm_60_days_after_ecg \
    wbc_scaled_60_days_after_ecg \
    weightkg_scaled_60_days_after_ecg \
--explore_stratify_label sts_death_60_days_after_ecg \
--output_folder $HOME/dropbox/sts-ecg/explore/voltage-age-sex-metadata-filter-sts
```

### PR DESCRIPTION
**Pull request by StevenSong**
## Issues closed
Closes #1347
Closes #1366 

## Changelog
- [x] adds some logic to the implementation of ReduceLROnPlateau to also reduce the weight decay (https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/keras/callbacks.py#L2522)

## Checklist
- [x] The following are complete:
    - Added or revised tests if appropriate.
    - If new packages are installed, `requirements.txt` is updated.
    - If documentation should be updated, a new issue was made in `ml4c3-wiki`.
    - The PR name describes its content and does not contain "Ref #".
    - The PR does not add PHI.

### CODE REVIEW

File: ml4c3/models.py [hunk 1]
```diff
@@ -1579,14 +1628,19 @@ def _get_callbacks(
         EarlyStopping(monitor=monitor, patience=patience, verbose=1),
     ]
     if learning_rate_schedule is None:
+        if optimizer == "adamw":
+            rlrop = ReduceLROnPlateauWithWD
```
Comment (erikr) [inline ml4c3/models.py:1632]: Can we use a more verbose name than `rlrop`?

File: ml4c3/models.py [hunk 2]
```diff
@@ -1561,12 +1563,59 @@ def _make_multimodal_multitask_model(
     return Model(inputs=inputs, outputs=list(decoder_outputs.values()))
 
 
+class ReduceLROnPlateauWithWD(ReduceLROnPlateau):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def on_epoch_end(self, epoch, logs=None):
+        logs = logs or {}
```
Comment (erikr) [inline ml4c3/models.py:1571]: If only keys in `logs` are `lr` and `weight_decay`, can we use a more descriptive name such as `optimizer_attributes`? `logs` is general and similar to `logging`. I think making the name descriptive and distinct provides helpful contrast.

File: ml4c3/models.py [hunk 3]
```diff
@@ -1561,12 +1563,59 @@ def _make_multimodal_multitask_model(
...
+
+            if self.monitor_op(current, self.best):
+                self.best = current
+                self.wait = 0
+            elif not self.in_cooldown():
+                self.wait += 1
+                if self.wait >= self.patience:
+                    old_wd = float(K.get_value(self.model.optimizer.weight_decay))
+                    old_lr = float(K.get_value(self.model.optimizer.lr))
+                    if old_lr > self.min_lr:
```
Comment (erikr) [inline ml4c3/models.py:1596]: Should we make the LR not change if it is greater than the minimum LR? Or, we could set the LR equal to `self.min_lr`. I don't have a strong preference either way, not sure what papers do, and not sure it really affects the final result. Can be future discussion & PR. Just a thought.

**Review by erikr — APPROVED** [review]
Awesome work! Minor nits, NNTBB.

File: ml4c3/models.py (hunk 2 above)
Comment (StevenSong) [reply ml4c3/models.py:1571]: can't change `logs` otherwise the function signature doesn't match the super class

Comment (erikr) [reply ml4c3/models.py:1571]: I now see you're following the syntax norms at https://github.com/tensorflow/tensorflow/blob/v2.4.1/tensorflow/python/keras/callbacks.py#L2522. If you plan on opening a PR to TensorFlow, `logs` here is fine. But if not, perhaps we can make ours more descriptive.

File: ml4c3/models.py (hunk 3 above)
Comment (StevenSong) [reply ml4c3/models.py:1596]: yes, good idea, will change

File: ml4c3/models.py (hunk 2 above)
Comment (StevenSong) [reply ml4c3/models.py:1571]: I suppose maybe I ought to send this back to TFA folks but regardless, the `logs` object can't really change because this function is called elsewhere in tensorflow and if there's ever a place that uses the key-word arg `logs`, our version will break

Comment (erikr) [reply ml4c3/models.py:1571]: Understood

File: ml4c3/models.py (hunk 3 above)
Comment (StevenSong) [reply ml4c3/models.py:1596]: ah sorry, I understand the suggestion now, the implementation in tensorflow is the latter, if the new learning rate would be below the minimum, the learning rate is set to the minimum

**Comment by StevenSong** [comment]
@erikr sorry i pushed the explore fix to this PR too, can you take a look?

File: ml4c3/explorations.py [hunk 4]
```diff
@@ -1020,7 +1022,7 @@ def _tensors_to_df_with_dataset(
         patient_ids=patient_ids,
         input_tmaps=tensor_maps_in,
         output_tmaps=[],
-        filter_tmaps=[],
+        filter_tmaps=tensor_maps_filter,
```
Comment (erikr) [inline ml4c3/explorations.py:1025]: In a future issue & PR perhaps we should make this nomenclature consistent, e.g. `input_tmaps` should instead be `tensor_maps_in`.

File: ml4c3/explorations.py [hunk 5]
```diff
@@ -89,6 +89,7 @@ def explore(
             patient_csv=args.patient_csv,
             mrn_column_name=args.mrn_column_name,
             tensor_maps_in=required_tmaps,
+            tensor_maps_filter=args.tensor_maps_out + args.tensor_maps_filter,
```
Comment (erikr) [inline ml4c3/explorations.py:92]: looks good

**Review by erikr — APPROVED** [review]
I reviewed, it looks good! One comment but no other to-do's IMO.
